waku/v2/waku_archive/driver/postgres_driver/postgres_driver.nim

```diff
@@ -0,0 +1,288 @@
+  var statements: seq[string]
+  var args: seq[string]
+
+  if contentTopic.len > 0:
+    let cstmt = "contentTopic IN (" & "?".repeat(contentTopic.len).join(",") & ")"
+    statements.add(cstmt)
+    for t in contentTopic:
+      args.add(t)
+
+  if pubsubTopic.isSome():
+    statements.add("pubsubTopic = ?")
+    args.add(pubsubTopic.get())
+
+  if cursor.isSome():
+    let comp = if ascendingOrder: ">" else: "<"
+    statements.add("(storedAt, id) " & comp & " (?,?)")
+    args.add($cursor.get().storeTime)
+    args.add($cursor.get().digest.data)
+
+  if startTime.isSome():
+    statements.add("storedAt >= ?")
+    args.add($startTime.get())
+
+  if endTime.isSome():
+    statements.add("storedAt <= ?")
+    args.add($endTime.get())
+
+  if statements.len > 0:
+    query &= " WHERE " & statements.join(" AND ")
+
+  var direction: string
+  if ascendingOrder:
+    direction = "ASC"
+  else:
+    direction = "DESC"
+
+  query &= " ORDER BY storedAt " & direction & ", id " & direction
+
+  query &= " LIMIT ?"
+  args.add($maxPageSize)
+
+  var rows: seq[Row]
+  var results: seq[ArchiveRow]
+  try:
+    rows = s.connection.getAllRows(sql(query), args)
+  except DbError:
+    return err("failed to query rows")
+
+  for r in rows:
+    let rowRes = extractRow(r)
+    if rowRes.isErr():
+      return err("failed to extract row")
+
+    results.add(rowRes.get())
+
+  return ok(results)
+
+method getMessagesCount*(s: PostgresDriver):
+                         Future[ArchiveDriverResult[int64]] {.async.} =
+  var count: int64
+  try:
+    let row = s.connection.getRow(sql("""SELECT COUNT(1) FROM messages"""))
+    count = parseInt(row[0])
+
+  except DbError:
+    return err("failed to query count")
+  except ValueError:
+    return err("failed to parse query count result")
+
+  return ok(count)
+
+method getOldestMessageTimestamp*(s: PostgresDriver):
+                                  Future[ArchiveDriverResult[Timestamp]] {.async.} =
+  return err("not implemented")
+
+method getNewestMessageTimestamp*(s: PostgresDriver):
+                                  Future[ArchiveDriverResult[Timestamp]] {.async.} =
+  return err("not implemented")
+
+method deleteMessagesOlderThanTimestamp*(s: PostgresDriver,
+                                         ts: Timestamp):
+                                         Future[ArchiveDriverResult[void]] {.async.} =
+  return err("not implemented")
+
+method deleteOldestMessagesNotWithinLimit*(s: PostgresDriver,
+                                           limit: int):
+                                           Future[ArchiveDriverResult[void]] {.async.} =
+  return err("not implemented")
+
+method close*(s: PostgresDriver):
+              Future[ArchiveDriverResult[void]] {.async.} =
+  ## Close the database connection
+  s.connection.close()
+  return ok()
+
+method sleep*(s: PostgresDriver, seconds: int):
+              Future[ArchiveDriverResult[void]] {.base, async.} =
```

Comment:
If this is not overridden/inherited anywhere else, it should not be annotated `base` and be a `proc` rather than a `method`.